### DISCUSSION TITLE
Custom claims no longer included in introspection response after upgrade (21 to 24)

### DISCUSSION THREAD

**Discussion by rmaidment** [keycloak/keycloak#31520] [Q&A]
Hello,

I hope someone can help with this.  After upgrading from v21 to v24, claims added by my custom protocol mapper are no longer included in the introspection endpoint response.  
The UI for my custom protocol mapper has only ever displayed two options (both enabled):
Add to access token
Add to ID token

Why the change in behavior after upgrade? Maybe I need to expose an "Add to introspection" option, and there was a bug in v21 that meant custom claims were included in the introspection response even when this option was not enabled, and this bug has been fixed in v24?

Thanks in advance

**Comment by cgeorgilakis** [comment]
Yes, you need to expose an "Add to introspection" option and make it true in order to work as you want. Only if this flag is true, mapper will be added to token introspection.
"Add to introspection" option was an addition in Keycloak in order to support a lightweight access token and enhancing token introspection response.
For more info see [lightweight access token documentation](https://www.keycloak.org/docs/latest/server_admin/index.html#_using_lightweight_access_token)

**Reply by rmaidment** [reply]
Thank you for this.  Why the change in behavior on upgrade though?  The [v23 release notes](https://www.keycloak.org/2023/11/keycloak-2300-released) say:

"When migrating from previous versions, the introspection endpoint should return same claims, which are returned from access token, so the behavior should be effectively the same by default after the migration."

**Reply by cgeorgilakis** [reply]
> Thank you for this. Why the change in behavior on upgrade though? The [v23 release notes](https://www.keycloak.org/2023/11/keycloak-2300-released) say:
> 
> "When migrating from previous versions, the introspection endpoint should return same claims, which are returned from access token, so the behavior should be effectively the same by default after the migration."

This is true for Keycloak mappers.
I do not know for extensions (custom mappers). Maybe Keycloak team can help.
